### ISSUE TITLE
Double serialization precision is not IEEE 754-compliant

### ISSUE THREAD

**Issue by cgnx-gdobra** [Tencent/rapidjson#2236]
**Repro:**

```c++
  double d = 0.91872198684135509;
  rapidjson::Document document;
  document.SetDouble(d);

  std::ostringstream jsonString;
  rapidjson::OStreamWrapper outputStreamWrapper(jsonString);
  rapidjson::Writer<rapidjson::OStreamWrapper> writer(outputStreamWrapper);
  document.Accept(writer);

  rapidjson::Document document2;
  document2.Parse(jsonString.str().c_str());
  double d2 = document2.GetDouble();
  EXPECT_TRUE(d2 == d);
```

The JSON contains 0.9187219868413551 and it is parsed back to d2 as 0.91872198684135520.

**Rationale:**

https://en.wikipedia.org/wiki/IEEE_754#Character_representation

>  Conversion to an external character sequence must be such that conversion back using round to nearest, ties to even will recover the original number. [...] The original binary value will be preserved by converting to decimal and back again using [...] 17 decimal digits for binary64.

**Comment by SciCed** [comment]
I haven't tried it with the DOM interface, but I'm using the parsing flag `kParseFullPrecisionFlag` with the SAX interface to fix the precision issue.

`document2.Parse<kParseFullPrecisionFlag>(jsonString.str().c_str());` may work in your case.

**Comment by cvanwinkle** [comment]
> `Parse<kParseFullPrecisionFlag>(...);` may work in your case.

Yes, this looks to be correct. I was tracking down issues where parsing wasn't the same on Posix systems vs Win, and it came down to this, just setting the appropriate flag on parse. Without it, successive `float64 <--> JSON` serializations could creep up to 7 ULP, and even toggle the LSB on each successive generation for other values. Setting `kParseFullPrecisionFlag` makes it roundtrip cleanly and consistently.
